srml/executive/src/lib.rs

```diff
@@ -580,6 +579,131 @@ mod tests {
 		run_test(true);
 	}
 
+	#[test]
+	fn block_weight_multiple_checks() {
+
+		let run_test = |overflow_tx: i8| {
```

Comment:
again, good test scenario but
1- Dependent on hardcoded values (could also be a test-local variable)
2- way too verbose. Use patterns like `let exts = (1..10).map(|i| primitives::testing::TestXt(Some(1), 2, Call::transfer(33, 69))).collect::<Vec<Extrinsic>>()` and then `exts.iter().take(i).for_each(Executive::apply_extrinsic)` 

(just a recommendation, not necessarily the best way imo)

Reply:
The basic idea here is to add a bunch of transactions and then test what happens when a *tipping* transaction is added that pushes the weight above the maximum weight. 

I did try to follow the pattern you recommended, but I don't think it works because it is actually testing five separate scenarios in which 5 permutations are tested where the last transaction pushes the weight above the limit. I spent a while trying to do it in the same `new_test_ext` and I don't think that's possible without changing what is being tested.

```rust
		let mut t = new_test_ext();
		// vector of tuples (index, transaction)
		let transactions: Vec<_> = (0..5).map(|t| {
			(t, primitives::testing::TestXt(Some(1), t, Call::transfer(33, 69)))
		}).collect();
		let mut counter = 0;
		transactions.iter().for_each(|i| {
			println!("c1: {}", counter);
			let encoded = i.1.encode();
			let len = (MAX_TRANSACTIONS_WEIGHT + 1) as usize;
			with_externalities(&mut t, || {
				Executive::initialize_block(&Header::new(1, H256::default(), H256::default(),
													[69u8; 32].into(), Digest::default()));
				assert_eq!(<system::Module<Runtime>>::all_extrinsics_weight(), 0);

				// for the purpose of adding previous transactions
				for (index, utx) in transactions.clone().iter() {
					if (index >= &0u64) && (index < &(counter as u64)) {
						let _ = Executive::apply_extrinsic(utx.clone()).unwrap();
						println!("total weight: {}", <system::Module<Runtime>>::all_extrinsics_weight());
					}
				}
				let res = Executive::apply_extrinsic_with_len(i.1.clone(), len, Some(encoded));
				assert!(res.is_err());
				println!("c2: {}", counter);
				assert_eq!(<system::Module<Runtime>>::all_extrinsics_weight(), (counter * 28) as u32);
				assert_eq!(<system::Module<Runtime>>::extrinsic_index(), Some(counter as u32));
			});
			counter += 1;
		});
```

Reply:
The difficult part is expressing conditional paths with just one invocation of `with_externalities` -- before, I was testing 5 different scenarios, but now I am limited to one scenario (there is no reset for `with_externalities`)

Reply:
> The basic idea here is to add a bunch of transactions and then test what happens when a tipping transaction is added that pushes the weight above the maximum weight.

This seems great + simple enough for one test-case. I'd say stick to doing this and only this. The 5 scenarios is too complicated for one unit test.